### PR TITLE
fix: single character compatibility with Python binding

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: python/pylibvw.cc [hunk 1]
```diff
@@ -546,31 +546,101 @@ std::string varray_char_to_string(v_array<char>& a)
 
 std::string my_get_tag(example_ptr ec) { return varray_char_to_string(ec->tag); }
 
-uint32_t ex_num_namespaces(example_ptr ec) { return (uint32_t)ec->indices.size(); }
+uint32_t ex_num_namespaces(example_ptr ec) { return ec->feature_space.end() - ec->feature_space.begin(); }
```
Comment (jackgerrits) [inline python/pylibvw.cc]: This does not need to change.
```suggestion
uint32_t ex_num_namespaces(example_ptr ec) { return (uint32_t)ec->indices.size(); }
```

File: python/pylibvw.cc [hunk 2]
```diff
@@ -546,31 +546,101 @@ std::string varray_char_to_string(v_array<char>& a)
...
 
-unsigned char ex_namespace(example_ptr ec, uint32_t ns) { return ec->indices[ns]; }
+namespace_index ex_namespace(example_ptr ec, uint32_t distance)
+{
+  auto it = ec->feature_space.begin();
+  for (int i = 0; i < distance; ++i) 
+  {
+    ++it;
+  }
+   return (*it).front().index;
```
Comment (jackgerrits) [inline python/pylibvw.cc]:
```suggestion
  return ec->feature_space.indices()[ns];
```

File: python/pylibvw.cc [hunk 3]
```diff
@@ -546,31 +546,101 @@ std::string varray_char_to_string(v_array<char>& a)
...
+  }
+  return ret;
+}
 
-uint32_t ex_feature(example_ptr ec, unsigned char ns, uint32_t i)
+uint32_t ex_feature(example_ptr ec, namespace_index idx, uint32_t i)
 {
-  return (uint32_t)ec->feature_space[ns].indicies[i];
+  auto& fs = ec->feature_space;
+  const auto& group_list = fs.get_list(idx);
```
Comment (jackgerrits) [inline python/pylibvw.cc]: You should use `index_flat_begin` instead of doing this manually

File: python/pylibvw.cc [hunk 4]
```diff
@@ -546,31 +546,101 @@ std::string varray_char_to_string(v_array<char>& a)
...
+      curr += nsf.feats.size();
+    }
+  }
+  THROW("Error: Index outside range of features");
 }
 
-float ex_feature_weight(example_ptr ec, unsigned char ns, uint32_t i) { return ec->feature_space[ns].values[i]; }
+float ex_feature_weight(example_ptr ec, namespace_index idx, uint32_t i)
+{
+  auto& fs = ec->feature_space;
```
Comment (jackgerrits) [inline python/pylibvw.cc:574]: You should use `index_flat_begin` instead of doing this manually

File: python/pylibvw.cc [hunk 5]
```diff
@@ -546,31 +546,101 @@ std::string varray_char_to_string(v_array<char>& a)
...
+  return ret;
+}
+
+void ex_push_namespace(example_ptr ec, namespace_index idx)
+{
+  auto& fs = ec->feature_space;
+  if (fs.get_list(idx).size() == 0)
+  {
+    char ns_str[2] = {(char)idx, 0};
+    uint64_t hash = hashstring(ns_str, 1, 0);
```
Comment (jackgerrits) [inline python/pylibvw.cc]: This must use VW::hash_space

File: python/pylibvw.cc [hunk 6]
```diff
@@ -712,30 +810,39 @@ void unsetup_example(vw_ptr vwP, example_ptr ae)
 
   if (all.add_constant)
   {
-    ae->feature_space[constant_namespace].clear();
-    int hit_constant = -1;
-    size_t N = ae->indices.size();
-    for (size_t i = 0; i < N; i++)
+    std::vector<unsigned char> bad_indices;
```
Comment (jackgerrits) [inline python/pylibvw.cc]: All you need to do is remove the constant namespace. No other feature groups are affected

Comment (bassmang) [reply python/pylibvw.cc]: All these changes are addressed in the latest commit

**Pull request merged**